library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java

```diff
@@ -18,324 +18,718 @@
+      windowDataManagers[0].deletedOperators = removedOperatorIds;
     }
+
+    List<WindowDataManager> mangers = new ArrayList<>();
+    mangers.addAll(Arrays.asList(windowDataManagers));
+    return mangers;
   }
 
   @Override
   public void teardown()
   {
-    try {
-      fs.close();
-    } catch (IOException e) {
-      throw new RuntimeException(e);
+    wal.teardown();
+    for (FSWindowReplayWAL wal : readOnlyWals.values()) {
+      wal.teardown();
+    }
+  }
+
+  protected void setRelyOnCheckpoints(boolean relyOnCheckpoints)
+  {
+    this.relyOnCheckpoints = relyOnCheckpoints;
+  }
+  
+  /**
+   * @return wal instance
+   */
+  protected FSWindowReplayWAL getWal()
+  {
+    return wal;
+  }
+  
+  @VisibleForTesting
+  public Set<Integer> getDeletedOperators()
+  {
+    if (deletedOperators == null) {
+      return null;
     }
+    return ImmutableSet.copyOf(deletedOperators);
   }
 
   /**
-   * @return recovery path
+   * @return recovery filePath
    */
-  public String getRecoveryPath()
+  public String getStatePath()
   {
-    return recoveryPath;
+    return statePath;
   }
 
   /**
-   * Sets the recovery path. If {@link #isRecoveryPathRelativeToAppPath} is true then this path is handled relative
-   * to the application path; otherwise it is handled as an absolute path.
+   * Sets the state path. If {@link #isStatePathRelativeToAppPath} is true then this filePath is handled
+   * relative
+   * to the application filePath; otherwise it is handled as an absolute filePath.
    *
-   * @param recoveryPath recovery path
+   * @param statePath recovery filePath
    */
-  public void setRecoveryPath(String recoveryPath)
+  public void setStatePath(String statePath)
   {
-    this.recoveryPath = recoveryPath;
+    this.statePath = statePath;
   }
 
   /**
-   * @return true if recovery path is relative to app path; false otherwise.
+   * @return true if state path is relative to app path; false otherwise.
    */
-  public boolean isRecoveryPathRelativeToAppPath()
+  public boolean isStatePathRelativeToAppPath()
   {
-    return isRecoveryPathRelativeToAppPath;
+    return isStatePathRelativeToAppPath;
   }
 
   /**
-   * Specifies whether the recovery path is relative to application path.
+   * Specifies whether the state path is relative to application filePath.
    *
-   * @param recoveryPathRelativeToAppPath true if recovery path is relative to application path; false otherwise.
+   * @param statePathRelativeToAppPath true if state path is relative to application path; false
+   *                                      otherwise.
    */
-  public void setRecoveryPathRelativeToAppPath(boolean recoveryPathRelativeToAppPath)
+  public void setStatePathRelativeToAppPath(boolean statePathRelativeToAppPath)
   {
-    isRecoveryPathRelativeToAppPath = recoveryPathRelativeToAppPath;
+    isStatePathRelativeToAppPath = statePathRelativeToAppPath;
   }
+
+  protected static Slice DELETED = new Slice("DEL".getBytes());
```

Comment:
It was just an effort made by JMS Input to delete the last saved artifact because it ack failed. It didn't rely on that call to be successful. If we add an end marker for window then we need to add an additional method to WindowDataManager api and every operator needs to invoke it after saving a window.